lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -569,6 +576,93 @@ newTransactionLayer networkId = TransactionLayer
 _decodeSealedTx :: SealedTx -> (Tx, TokenMap, TokenMap, [Certificate])
 _decodeSealedTx (cardanoTx -> InAnyCardanoEra _era tx) = fromCardanoTx tx
 
+_evaluateTransactionBalance
+    :: SealedTx
+    -> Cardano.ProtocolParameters
+    -> UTxO
+    -> [(TxIn, TxOut, Maybe (Hash "Datum"))]
+    -> Maybe Cardano.Value
+_evaluateTransactionBalance tx pp utxo extraUTxO = do
+    shelleyTx <- inAnyShelleyBasedEra $ cardanoTx tx
+    pure $ withShelleyBasedBody shelleyTx $ \era bod ->
+        let
+            utxo' = Map.fromList
+                . map (bimap toCardanoTxIn (toCardanoTxOut era))
+                . Map.toList
+                $ unUTxO utxo
+
+            extraUTxO' = Map.fromList
+                . map (\(i, o, mDatumHash) ->
+                    (toCardanoTxIn i, setDatumHash era mDatumHash (toCardanoTxOut era o))
+                    )
+                $ extraUTxO
+
+        in
+            lovelaceFromCardanoTxOutValue
+                $ Cardano.evaluateTransactionBalance
+                    pp
+                    mempty
+                    (Cardano.UTxO $ utxo' <> extraUTxO')
+                    -- NOTE: We don't want the keys to overlap! Unclear how to
+                    -- address.
+                    bod
+  where
+    setDatumHash :: ShelleyBasedEra era -> Maybe (Hash "Datum") -> Cardano.TxOut ctx era -> Cardano.TxOut ctx era
+    setDatumHash _era Nothing o = o
+    setDatumHash era (Just (Hash datumHash)) (Cardano.TxOut addr val _) =
+        Cardano.TxOut addr val (Cardano.TxOutDatumHash scriptDataSupported hash)
+      where
+        scriptDataSupported = case era of
+            ShelleyBasedEraAlonzo -> Cardano.ScriptDataInAlonzoEra
+            ShelleyBasedEraMary -> errBadEra
+            ShelleyBasedEraAllegra -> errBadEra
+            ShelleyBasedEraShelley -> errBadEra
+          where
+            -- FIXME: Proper error handling
+            errBadEra = error $ unwords
+                    [ "evaluateTransactionBalance:"
+                    , "cannot add a datum hash to the transaction body of an"
+                    , "era which doesn't support datum hashes."
+                    ]
```

Comment:
I thought I'd address these, but TBH, because there probably are higher priority problems, it might actually make sense to leave these for now...